### PR TITLE
MLPerf HPC Weak scaling metric proposal

### PR DESCRIPTION
**Pull request by azrael417**
Sear Sir/Madam,

this PR contains the rules changes for MLperf HPC concerning the updated performance metrics. Please consider this draft WIP as some details are still being discussed. As such, I encourage members of the MLPerf HPC WG to comment on the PR and refine it before it can ultimately be merged.

Best regards
Thorsten

### CODE REVIEW

File: hpc_training_rules.adoc [hunk 1]
```diff
@@ -92,16 +92,50 @@ OPEN: Hyperparameters and optimizer may be freely changed.
...
+=== Strong Scaling (Time to Convergence)
+This is a *mandatory* metric: see MLPerf Training xref:training_rules.adoc#section-run-results[Rule 11] for reference. The same rules apply here.
+
+=== Data staging
+This is a *mandatory* metric: staging of data from remote distributed storage to node-local storage or I/O accelerators must be timed as described in <<Data State at Start of Run>> above.
+
+=== Weak Scaling (Throughput)
+This is an *optional* metric. It was designed to test the training capacity of a system.
+
+Measurement: the submitter picks a number of independent training instances K and machine scale S. This scale can be defined as:
```
Comment (sparticlesteve) [inline hpc_training_rules.adoc]: I think it's better if S (or S') represents the scale of each training instance. Then the total system size used is S*K. I think it's more relatable to the traditional metric reporting, which is similarly defined by training scale S.

File: hpc_training_rules.adoc [hunk 2]
```diff
@@ -92,16 +92,50 @@ OPEN: Hyperparameters and optimizer may be freely changed.
...
+* number of nodes
+* number of accelerators (e.g. GPU)
+* number of canonical compute units for a given system
```
Comment (sparticlesteve) [inline hpc_training_rules.adoc]: you could shorten this; something like "S is the number of compute units (e.g. accelerators, nodes) for a given system."

File: hpc_training_rules.adoc [hunk 3]
```diff
@@ -92,16 +92,50 @@ OPEN: Hyperparameters and optimizer may be freely changed.
...
+* The submitter *must not report this score on it's own*. It has to be reported in conjunction with the scores from <<Strong Scaling (Time to Convergence)>> and <<Data staging>>.
+* this score *does not allow for extrapolation*. All K training instances have to be run to convergence at chosen scale S.
+
 
-You are encouraged to also report the additional metrics of the benchmark described in the MLPerf HPC v0.5 Metrics List (to be prepared).
 
 == Benchmark Results
 
 We follow the MLPerf Training Rule 11 along with the following required number of runs per benchmark.
+Note that since run-to-run variability is already captured by spatial multiplexing in case of metric 3, the number of required runs can be reduced. More precisely, the submitter has to obey the following rule:
```
Comment (sparticlesteve) [inline hpc_training_rules.adoc]: I'm a little confused by "the number of required runs can be reduced". Can you explain?

File: hpc_training_rules.adoc [hunk 4]
```diff
@@ -92,16 +92,50 @@ OPEN: Hyperparameters and optimizer may be freely changed.
...
 
 |===
-|Benchmark |Number of Runs
-|DeepCAM | 5
-|CosmoFlow | 10
+|Benchmark |Number of Runs (Metric 1, 2) | Instance Count x Number of Runs (Metric 3)
```
Comment (sparticlesteve) [inline hpc_training_rules.adoc]: Why is it (Instance Count x Number of Runs) ?
Instance Count is "K", right? I thought we only ran K models for metric 3.

Should it just say "Number of concurrent runs (Metric 3)", i.e. "K", or something like that?

**Review by sparticlesteve — COMMENTED** [review]
Thanks for the PR, Thorsten. I added some comments.

Are you already covered by CLA? Do we just need to get your github account authorized?

**Comment by sparticlesteve** [comment]
Copying some of my comments here from slack, for posterity:

Using max(end_time) - min(start_time) will mean the result is entirely driven by the slowest model, and is thus highly sensitive to outliers (i.e. bad luck) in training variability. This is of course why we use olympic scoring in the traditional metric reporting. A possible way to mitigate this is to run K models, but drop the last and compute the time using just the K-1 runs.

David had a further suggestion, which I think was along the lines of using the average runtime across the K concurrent runs. This is more resilient to variability. However, I think it would require us to enforce that all jobs actually run at the same time.

**Comment by sparticlesteve** [comment]
More discussion points copied from slack, for posterity:

* How exactly do we define a “system”? What level of heterogeneity (partitions with different hardware) do we allow? Do we allow system partitions to be spread out geographically (probably not useful)?
* Should we impose a minimum (or fixed) batch size? This would allow us some control over runtime and disincentivizing folks from running smallest possible batch size for longest possible system run time (which disadvantages folks who cannot get that amount of system access)

File: hpc_training_rules.adoc (hunk 1 above)
Comment (azrael417) [reply hpc_training_rules.adoc]: that is fine with me

File: hpc_training_rules.adoc (hunk 3 above)
Comment (azrael417) [reply hpc_training_rules.adoc]: that means we still want to see 10 runs for cosmoflow and 5 runs for deepcam but since we use spatial multiplexing we only require that K*num_runs >= [5|10]. Does that make sense?

File: hpc_training_rules.adoc (hunk 4 above)
Comment (azrael417) [reply hpc_training_rules.adoc]: See above. We want to make sure we still have at least 5 or 10 independent runs respectively. It is very easy to fulfill this requirement so it shouldn't be a big discussion point.

**Comment by azrael417** [comment]
How exactly do we define a “system”? What level of heterogeneity (partitions with different hardware) do we allow? Do we allow system partitions to be spread out geographically (probably not useful)?

Actually I think this is irrelevant for this metric, since you have to run at the scale you report. So if you can run on all your cloud instances together that should be fine. It is unlikely that someone is going to do that anyway. That means we probably do not need to define what a system is.

**Comment by azrael417** [comment]
Should we impose a minimum (or fixed) batch size? This would allow us some control over runtime and disincentivizing folks from running smallest possible batch size for longest possible system run time (which disadvantages folks who cannot get that amount of system access)

No, imo that should be open. You can also get a large minimum batch size on a single accelerator if you use gradient accumulation for example. I think we should just leave that open.

File: hpc_training_rules.adoc [hunk 5]
```diff
@@ -104,29 +104,35 @@ This is a *mandatory* metric: staging of data from remote distributed storage to
...
 
-We will use the term compute unit as a generic term. Note that S can be different from those used for submitting metric 1. 
-Ideally S is the scale of the full system but this is not required.  
+* number of models M: number of model instances which are going to be trained in this benchmark.
+* instance scale S: each individual model instance will be trained at this scale. 
+* total utilized scale T: the total scale used for running this benchmark. For example, if all M models are trained concurrently, then T=M*S. More generally we can write that S<=T<=M*S if (some of) the models are trained sequentially.
 
-Each training instance will cover S/K compute units. The submitter trains these K instances to convergence. 
+Notes:
+* All three numbers M,S,T are chosen by the submitter. This allows the submnitter to accomodate his submission to available machine resources, i.e. compute capacity and compute time.
```
Comment (sparticlesteve) [inline hpc_training_rules.adoc]: submnitter -> submitter

his -> their

File: hpc_training_rules.adoc [hunk 6]
```diff
@@ -104,29 +104,35 @@ This is a *mandatory* metric: staging of data from remote distributed storage to
...
 
 We define a Time-To-Train-all (TTTa) number by computing the difference between the end time of the instance which needs longest time to converge and the start time of the instance which starts up fastest. Mathematically this can be expressed as 
 
 ----
-TTTa = max(run_stop) - min(run_start) where the max/min are taken over all instances K. 
+TTTa = max(run_stop) - min(run_start) where the max/min are taken over all instances M. 
 ----
 
-Reporting: the submitter reports the the tuple (K, TTTa). 
+Note: the submitter is allowed to prune this number by removing results from individual training instances. As long as the minimum number of models rule is satisfied (see section <<Benchmark Results>> below), the submission is valid. He then uses a modified number of models M'<=M and computes TTTa over the reduced set. This allows the submitter to remove occasional outliers or stragglers which would otherwise reduce the score disproportionally.
```
Comment (sparticlesteve) [inline hpc_training_rules.adoc]: He -> They or "The submitter"

File: hpc_training_rules.adoc [hunk 7]
```diff
@@ -104,29 +104,35 @@ This is a *mandatory* metric: staging of data from remote distributed storage to
...
+Measurement: we will define 3 important parameters first. 
 
-We will use the term compute unit as a generic term. Note that S can be different from those used for submitting metric 1. 
-Ideally S is the scale of the full system but this is not required.  
+* number of models M: number of model instances which are going to be trained in this benchmark.
+* instance scale S: each individual model instance will be trained at this scale. 
+* total utilized scale T: the total scale used for running this benchmark. For example, if all M models are trained concurrently, then T=M*S. More generally we can write that S<=T<=M*S if (some of) the models are trained sequentially.
 
-Each training instance will cover S/K compute units. The submitter trains these K instances to convergence. 
+Notes:
```
Comment (sparticlesteve) [inline hpc_training_rules.adoc:109]: this list doesn't get resolved as bullets. Maybe it needs a blank line after this.

File: hpc_training_rules.adoc (hunk 2 above)
Comment (azrael417) [reply hpc_training_rules.adoc]: fixed

File: hpc_training_rules.adoc (hunk 3 above)
Comment (azrael417) [reply hpc_training_rules.adoc]: fixed

File: hpc_training_rules.adoc (hunk 5 above)
Comment (azrael417) [reply hpc_training_rules.adoc]: fixed, please review

File: hpc_training_rules.adoc (hunk 6 above)
Comment (azrael417) [reply hpc_training_rules.adoc]: fixed, please review

File: hpc_training_rules.adoc (hunk 7 above)
Comment (azrael417) [reply hpc_training_rules.adoc:109]: should also be fixed now

**Review by sparticlesteve — APPROVED** [review]
LGTM

File: hpc_training_rules.adoc [hunk 8]
```diff
@@ -92,16 +92,53 @@ OPEN: Hyperparameters and optimizer may be freely changed.
...
+Restrictions: 
+
+* The submitter *must not report this score on its own*. It has to be reported in conjunction with at least one score from <<Strong Scaling (Time to Convergence)>> from the same benchmark.
+* this score *does not allow for extrapolation*. All reported M' training instances must have converged and it is not allowed to extrapolate results in S or T.
 
-You are encouraged to also report the additional metrics of the benchmark described in the MLPerf HPC v0.5 Metrics List (to be prepared).
 
 == Benchmark Results
 
 We follow the MLPerf Training Rule 11 along with the following required number of runs per benchmark.
```
Comment (matthew-frank) [inline hpc_training_rules.adoc:137]: The section numbering in the other document seems to have changed since this sentence was originally written, so the Benchmark Results section over there is now "Rule 12" not "Rule 11".  Would it work to change this to [MLPerf Training Benchmark Results Rule](https://github.com/mlcommons/training_policies/blob/master/training_rules.adoc#benchmark-results)?

**Review by matthew-frank — APPROVED** [review]

File: hpc_training_rules.adoc (hunk 8 above)
Comment (sparticlesteve) [reply hpc_training_rules.adoc:137]: Thanks, yeah, I agree we should change that.

**Comment by sparticlesteve** [comment]
@johntran-nv could you take a look and merge?

**Comment by sparticlesteve** [comment]
@azrael417 can you remove WIP from the title? I think it is ready to be merged.
